[Clang] Reapply CWG2369 "Ordering between constraints and substitution"

zyn0217 opened the pull request:
The previous approach broke code generation for the MS ABI due to an unintended code path during constraint substitution. This time we address the issue by inspecting the evaluation contexts and thereby avoiding that code path.

This reapplies 96eced624 (#102857).

zyn0217 committed:
Reapply "[Clang] Implement CWG2369 "Ordering between constraints and substitution"" (#122130)

This reverts commit 3972ed57088f6515b787d7d38dec03dc74e51827.

zyn0217 committed:
Prevent dependent code generation when substituting constraints

zyn0217 committed:
Add a regression test

zyn0217 committed:
Switch to checking EvaluationContext

zyn0217 committed:
fixup! Switch to checking EvaluationContext

zyn0217 committed:
negative, sorry

zyn0217 committed:
Remove unused header

cor3ntin commented on clang/lib/Sema/SemaTemplateInstantiateDecl.cpp:
```diff
@@ -5284,8 +5284,22 @@ void Sema::InstantiateFunctionDefinition(SourceLocation PointOfInstantiation,
 
-  DeclGroupRef DG(Function);
-  Consumer.HandleTopLevelDecl(DG);
+  // With CWG2369, we substitute constraints before instantiating the associated
+  // function template. This helps prevent potential code generation for
+  // dependent types, particularly under the MS ABI.
```
```suggestion
  // We never need to emit the code for a lambda in unevaluated context.
  // We also can't mangle a lambda in the require clause of a function template 
  // during constraint checking as the MSI ABI would need to mangle the (not yet specialized) enclosing declaration
  // FIXME: Should we try to skip this for non-lambda functions too?
```

cor3ntin commented:
I've confirmed with @zyn0217 that this iteration of the patch fixes the bug encountered by the chromium team wherein we tried to mangle lambdas with a deduced return type during constraint satisfaction - which caused failure on windows as mangling of lambda requires the enclosing context to be mangled. 

---

However if we merge this PR, the bug described in https://github.com/llvm/llvm-project/issues/121881 
would be reintroduced.

The issue is that users seem to rely on a GCC specific behavior where GCC does not check template candidates during overload resolution if an exact match exists. 
This is a known divergence that is known to both Clang and WG21 (see #62096)

Given that GCC's behavior is strictly non-conforming but long standing, I have asked CWG for guidance 
https://lists.isocpp.org/core/2025/01/16814.php

That being said, I think our option are
 
- Wait for WG21 to decide what to do and implement whatever resolution they come up with (which may or not be a breaking change for us)

- Follow GCC's behavior immediately, however there are some drawbacks as it seems like users which relies on it do so inadvertently (and some diagnostics would not be emitted)
 
- Only implement this DR under pedantic or c++26 mode to limit breakage

- Accept the breakage  

@zygoloid @hubert-reinterpretcast @erichkeane @AaronBallman

zyn0217 commented:
@mysterymath @zmodem Can you both test this patch on Windows to confirm if it resolves issues you've reported? We'd greatly appreciate it!

alexfh commented:
FYI, we've also seen problems compiling code that includes certain boost headers (version 1.84, as far as I can tell) after https://github.com/llvm/llvm-project/pull/102857:
```
boost/asio/any_io_executor.hpp:161:13: error: satisfaction of constraint '::boost::asio::execution::executor<Executor>' depends on itself
  161 |   template <BOOST_ASIO_EXECUTION_EXECUTOR Executor>
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
...
```

I'm building clang with this PR to see if it resolves this issue.

zmodem commented:
Confirmed Chromium on Windows builds with this patch.

cor3ntin commented:
@alexfh It wouldn't fix your issue, see this comment https://github.com/llvm/llvm-project/pull/122423#issuecomment-2585390531 (there were 2 issues, a bug in the PR implementation, and an unfortunate breakage caused by some libraries relying on non-conforming behavior)

alexfh commented:
> @alexfh It wouldn't fix your issue, see this comment [#122423 (comment)](https://github.com/llvm/llvm-project/pull/122423#issuecomment-2585390531) (there were 2 issues, a bug in the PR implementation, and an unfortunate breakage caused by some libraries relying on non-conforming behavior)

Indeed, it doesn't fix this case. Neither does it fix another compilation error in the same version of boost:
```
boost/asio/detail/handler_work.hpp:452:7: error: ambiguous partial specializations of 'handler_work_base<boost::asio::any_io_executor>'
  452 |   > : handler_work_base<IoExecutor>
      |       ^
```

Endilll commented:
This was discussed during Clang C/C++ Language Working Group meeting today, and the consensus seems that this PR shouldn't land before Clang 20 branch, which is scheduled to happen on January 28th.

alexfh commented:
Can we have this behavior guarded by a compiler option?

hubert-reinterpretcast commented:
> Can we have this behavior guarded by a compiler option?

That depends on the rationale:
1. Allow opt in to the resolution of CWG2369 because it is wanted (despite potentially-temporary fallout); or
2. Require opt in to the resolution of CWG2369 because "it causes fallout".

For (2), we can simply delay enabling the resolution until the fallout is mitigated or accepted.

zyn0217 committed:
Implement GCC's CWG 2369 heuristic

zyn0217 commented on clang/lib/Sema/SemaTemplateInstantiateDecl.cpp:
```diff
@@ -5284,8 +5284,22 @@ void Sema::InstantiateFunctionDefinition(SourceLocation PointOfInstantiation,
+  // With CWG2369, we substitute constraints before instantiating the associated
+  // function template. This helps prevent potential code generation for
+  // dependent types, particularly under the MS ABI.
+  bool ShouldSkipCG = [&] {
+    auto *RD = dyn_cast<CXXRecordDecl>(Function->getParent());
+    if (!RD || !RD->isLambda())
+      return false;
+
+    return llvm::any_of(ExprEvalContexts, [](auto &Context) {
+      return Context.isUnevaluated() || Context.isImmediateFunctionContext();
+    });
+  }();
+  if (!ShouldSkipCG) {
+    DeclGroupRef DG(Function);
+    Consumer.HandleTopLevelDecl(DG);
+  }
```
Thanks, I'll take a look

zyn0217 committed:
Use UnsignedOrNone

cor3ntin commented on clang/lib/Sema/SemaOverload.cpp:
```diff
@@ -8058,6 +8059,54 @@ bool Sema::CheckNonDependentConversions(
     }
   }
 
+  // A heuristic & speculative workaround for bug
+  // https://gcc.gnu.org/bugzilla/show_bug.cgi?id=99599 that manifests after
+  // CWG2369.
```
I think that we should simplify that to just "has user-defined conversion operator" - I think both GCC folks and @zygoloid agree this is a simpler direction

zyn0217 replied:
Yes, I'll look into that; this is just rebasing from my old attempt

zyn0217 committed:
Check the non-user-defined-non-dependent conversions earlier

zyn0217 commented on clang/lib/Sema/SemaOverload.cpp:
```diff
@@ -8058,6 +8061,46 @@ bool Sema::CheckNonDependentConversions(
+      ArgType = ArgType->getPointeeType();
+    }
+
+    if (auto *RT = ParmType->getAs<RecordType>())
+      if (auto *RD = dyn_cast<CXXRecordDecl>(RT->getDecl());
+          RD && RD->hasDefinition()) {
+        if (llvm::any_of(LookupConstructors(RD), [](NamedDecl *ND) {
+              auto Info = getConstructorInfo(ND);
+              if (!Info)
+                return false;
+              CXXConstructorDecl *Ctor = Info.Constructor;
+              /// isConvertingConstructor takes copy/move constructors into
+              /// account!
+              return !Ctor->isCopyOrMoveConstructor() &&
+                     Ctor->isConvertingConstructor(
+                         /*AllowExplicit=*/true);
+            }))
+          return true;
+      }
```
For the `{foo,bar}_tag` issue, GCC still rejects it if the ParamType is not an aggregation:

```cpp

struct bar_tag {
   bar_tag() {} // user-defined constructor
};
struct foo_tag {
};

template <class T>
concept fooable = requires(T it) {
  invoke_tag(foo_tag{}, it); // <-- here
};

template <class T> auto invoke_tag(foo_tag, T in) { return in; }

template <fooable T> auto invoke_tag(bar_tag, T it) { return it; }

int main() {
  // Neither line below compiles in GCC 11, independently of the other
  return invoke_tag(foo_tag{}, 2) + invoke_tag(bar_tag{}, 2);
}

```

https://godbolt.org/z/bYcve4xqb

While checking user-defined conversions helps, we still end up with self-dependent diagnostics if `bar_tag` includes a conversion constructor:
```cpp
struct bar_tag {
   bar_tag(int) {}
};
```

cor3ntin replied:
GCC did the minimum not to break existing code. Because tag_invoke (and somewhat still is) fairly widely used.

In that light, I think it makes sense to just do as they did, or simpler (to the point it's unclear to me if we should even standardize that behavior)

zyn0217 committed:
Check only the aggregates, matching GCC's behavior

zyn0217 committed:
Fix comments

cor3ntin commented:
@zyn0217 is that ready for review?

zyn0217 committed:
Revert "Check only the aggregates, matching GCC's behavior"

This reverts commit 8423cefd2e4d123f5bd05b93c17eb46f186469e4.

zyn0217 committed:
More tests!

zyn0217 committed:
[Clang] Profile singly-resolved UnresolvedLookupExpr with the declaration

For a dependent variable template specialization, we don't build a
dependent Decl node or a DeclRefExpr to represent it. Instead, we
preserve the UnresolvedLookupExpr until instantiation.

However, this approach isn't ideal for constraint normalization. We
consider the qualifier during profiling, but since that's based on the
written code, it can introduce confusing differences, even when the
expressions resolve to the same declaration.

This change ensures that, if possible, we profile the resolved
declaration instead of its qualifier. For expressions that resolve
to more than one declarations, we still profile its qualifier, as
otherwise it would make us depend on the order of lookup results.

zyn0217 committed:
Merge branch 'main' into cwg-2369-again

zyn0217 committed:
Revert "[Clang] Profile singly-resolved UnresolvedLookupExpr with the declaration"

This reverts commit debef50c5cc472ded7e705d029dd351df5759ca7.

zyn0217 committed:
Fix param names

cor3ntin commented on clang/include/clang/Sema/Sema.h:
```diff
@@ -12554,7 +12555,9 @@ class Sema final : public SemaBase {
       SmallVectorImpl<OriginalCallArg> const *OriginalCallArgs,
       bool PartialOverloading, bool PartialOrdering,
-      llvm::function_ref<bool()> CheckNonDependent = [] { return false; });
+      llvm::function_ref<bool(bool)> CheckNonDependent = [](bool) {
+        return false;
+      });
```
Can you add comment to explain how the call back work / give a name to the parameter?

zyn0217 replied:
I added some comments and a param name

cor3ntin commented on clang/include/clang/Sema/Sema.h:
```diff
@@ -10506,7 +10506,8 @@ class Sema final : public SemaBase {
       FunctionTemplateDecl *FunctionTemplate, ArrayRef<QualType> ParamTypes,
       ArrayRef<Expr *> Args, OverloadCandidateSet &CandidateSet,
       ConversionSequenceList &Conversions, bool SuppressUserConversions,
-      CXXRecordDecl *ActingContext = nullptr, QualType ObjectType = QualType(),
+      bool SkipUserDefinedConversions, CXXRecordDecl *ActingContext = nullptr,
+      QualType ObjectType = QualType(),
```
SuppressUserConversions and SkipUserDefinedConversions is going to be super confusing. 
Ca=n we replace both booleans by a single enum, and add comments?

zyn0217 replied:
I packed them into a struct, along with some comments

cor3ntin commented on clang/include/clang/Sema/Sema.h:
```diff
@@ -13372,6 +13376,17 @@ class Sema final : public SemaBase {
       bool SkipForSpecialization = false,
       bool ForDefaultArgumentSubstitution = false);
 
+  /// Apart from storing the result to \p Result, this behaves the same as
+  /// another overload.
```
It would be cleaner to put the documentation on this overload, declare it first, and then say that the other overload creates a new MultiLevelTemplateArgumentList and call getTemplateInstantiationArgs with it as the first param)

cor3ntin commented on clang/include/clang/Sema/Template.h:
```diff
@@ -522,6 +522,12 @@ enum class TemplateSubstitutionKind : char {
+    /// Similar to \p findInstantiationOf(), but it wouldn't assert if the
+    /// instantiation was not found within the current instantiation scope. This
+    /// is helpful for on-demand declaration instantiation.
+    llvm::PointerUnion<Decl *, DeclArgumentPack *> *
+    findInstantiationUnsafe(const Decl *D);
```
`findInstantiationUnsafe` (can return null) and `getInstantiationOf` (asserts) would probably be better names.
Move that declaration _before_ the existing one.

cor3ntin commented on clang/lib/Sema/SemaTemplateInstantiate.cpp:
```diff
@@ -1921,9 +1956,62 @@ Decl *TemplateInstantiator::TransformDecl(SourceLocation Loc, Decl *D) {
     // template parameter.
   }
 
+  if (SemaRef.CurrentInstantiationScope) {
+    if (SemaRef.inConstraintSubstitution() && isa<ParmVarDecl>(D) &&
+        maybeInstantiateFunctionParameterToScope(cast<ParmVarDecl>(D)))
+      return nullptr;
+  }
```
You can merge the two ifs

cor3ntin commented on clang/lib/Sema/SemaTemplateInstantiate.cpp:
```diff
@@ -1432,12 +1459,20 @@ namespace {
+          SemaRef.inConstraintSubstitution()) {
+        for (UnexpandedParameterPack ParmPack : Unexpanded) {
+          NamedDecl *VD = ParmPack.first.dyn_cast<NamedDecl *>();
+          if (!isa_and_present<ParmVarDecl>(VD))
+            continue;
+          if (maybeInstantiateFunctionParameterToScope(cast<ParmVarDecl>(VD)))
+            return true;
```
You can use dyn_cast

cor3ntin commented on clang/include/clang/Sema/Sema.h:
```diff
@@ -13458,6 +13473,10 @@ class Sema final : public SemaBase {
   // FIXME: Should we have a similar limit for other forms of synthesis?
   unsigned NonInstantiationEntries;
 
+  /// The number of \p CodeSynthesisContexts that are not constraint
+  /// substitution.
+  unsigned NonConstraintSubstitutionEntries;
+
```
Why not use a bool flag that is inherited on construction?

cor3ntin reviewed (commented):
Thanks a lot for working on this

zyn0217 committed:
Fix a regression involving injected class types

zyn0217 committed:
checkpoint

zyn0217 committed:
Pack the two flags

cor3ntin reviewed (approved):
LGTM

zyn0217 committed:
Merge branch 'main' into cwg-2369-again

rupprecht commented:
I bisected a build failure to this commit. I can't figure out if it's expected or not. I minimized it to this:

```c++
#include <concepts>

enum class KindEnum {
    Unknown = 0,
    Foo = 1,
};

template <typename T>
concept KnownKind = T::kind() != KindEnum::Unknown;

template <KnownKind T>
struct KnownType;

struct Type {
    KindEnum kind() const;

    static Type f(Type t);

    template <KnownKind T>
    static KnownType<T> f(T t);

    static void g() {
        Type t;
        f(t);
    }
};

template <KnownKind T>
struct KnownType {
    static constexpr KindEnum kind() { return KindEnum::Foo; }
};
```

gcc accepts it. After this commit, clang rejects it:
```
<source>:9:21: error: substitution into constraint expression resulted in a non-constant expression
    9 | concept KnownKind = T::kind() != KindEnum::Unknown;
      |                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<source>:19:15: note: while checking the satisfaction of concept 'KnownKind<Type>' requested here
   19 |     template <KnownKind T>
      |               ^
<source>:19:15: note: while substituting template arguments into constraint expression here
   19 |     template <KnownKind T>
      |               ^~~~~~~~~
<source>:24:9: note: while checking constraint satisfaction for template 'f<Type>' required here
   24 |         f(t);
      |         ^
<source>:24:9: note: while substituting deduced template arguments into function template 'f' [with T = Type]
<source>:9:24: note: implicit use of 'this' pointer is only allowed within the evaluation of a call to a 'constexpr' member function
    9 | concept KnownKind = T::kind() != KindEnum::Unknown;
      |                        ^
1 error generated.
```

Live link: https://godbolt.org/z/oGh319a44

Is this sort of breakage expected?

zyn0217 commented:
```cpp

template <typename> class basic_ostream;
template <typename _Tp> constexpr bool is_trivial_v = __is_trivial(_Tp);
template <typename _CharT> struct basic_string_view {
  static_assert(__is_trivial(_CharT));
};
template <typename _CharT, typename>
void operator<<(basic_ostream<_CharT>, basic_string_view<_CharT>);
struct basic_ios {
  operator bool();
};
template <typename> struct basic_ostream : virtual basic_ios {};
template <typename _Ostream, typename _Tp> void operator<<(_Ostream, _Tp);
basic_ostream<char> g00_a0;
void g00() {
  __is_same(int, decltype(operator<< <basic_ostream<char>, char>(g00_a0, 1)));
}

```

So this is a issue brought by the heuristic non-dependent check which happens before constraint checking, where we think `basic_string_view<char>` won't entail a user-defined conversion and thus we instantiate it without checking the conversion `basic_ostream<char> -> basic_ostream<basic_ostream<char>>` first as we think that involves a user-defined conversion.

However, if we swapped the order of template parameters, GCC would suffer from the same issue too:

https://godbolt.org/z/M6xfGMPMo

```cpp

template <typename> class basic_ostream;

template <typename _CharT> struct basic_string_view {
  static_assert(__is_trivial(_CharT));
};

template <typename _CharT, typename>
void Foo(basic_string_view<_CharT>, basic_ostream<_CharT>);

struct basic_ios {
  operator bool();
};

template <typename> struct basic_ostream : virtual basic_ios {};

template <typename _Tp, typename _Ostream> void Foo(_Tp, _Ostream);

basic_ostream<char> g00_a0;

void g00() {
  static_assert(!__is_same(int, decltype(Foo<basic_ostream<char>, char>(g00_a0, 1))));
}

```

It's not quite clear to me if it is assured that no arbitrary template instantiation would occur during template argument deduction, so CC @zygoloid @cor3ntin @erichkeane for help :)

alexfh commented:
We've found another issue caused by this patch. I'm not quite sure what the standard says about this, but Clang after this commit is the only recent mainstream compiler that rejects this code (https://gcc.godbolt.org/z/KW56orYca):
```
struct basic_ostream {
  basic_ostream(int);
};
namespace n {
template <class R_> struct Point_2 : R_::Kernel_base::Point_2 {};
template <class R> void insert(basic_ostream, Point_2<R>, int);
struct S1 {};
struct S2 {};
} // namespace n
struct P {};
template <class>
void insert(n::S1 s1, n::S2 s2, int) {
  insert<P>(s1, s2, 0);
}
```

```
<source>:5:42: error: no member named 'Kernel_base' in 'P'
    5 | template <class R_> struct Point_2 : R_::Kernel_base::Point_2 {};
      |                                      ~~~~^
<source>:13:17: note: in instantiation of template class 'n::Point_2<P>' requested here
   13 |   insert<P>(s1, s2, 0);
      |                 ^
<source>:13:3: note: while substituting deduced template arguments into function template 'insert' [with R = P]
   13 |   insert<P>(s1, s2, 0);
      |   ^
1 error generated.
```

The test case above is reduced from https://github.com/CGAL/cgal/blob/5add7e715031a190a87cf6cdc9ad324543a1621a/Triangulation_3/test/Triangulation_3/test_regular_remove_3.cpp, where the error looks like this:
```
cgal/Kernel_23/include/CGAL/Point_2.h:32:28: error: no member named 'Kernel_base' in 'point_iterator_0'
   32 | class Point_2 : public R_::Kernel_base::Point_2
      |                        ~~~~^
cgal/Triangulation_3/test/Triangulation_3/test_regular_remove_3.cpp:350:34: note: in instantiation of template class 'CGAL::Point_2<point_iterator_0>' requested here
  350 |     insert<point_iterator_0> (T, points, 10);
      |                                  ^
cgal/Triangulation_3/test/Triangulation_3/test_regular_remove_3.cpp:350:5: note: while substituting deduced template arguments into function template 'insert' [with R = point_iterator_0]
  350 |     insert<point_iterator_0> (T, points, 10);
      |     ^
```

zygoloid commented:
A slightly modified version of that testcase gives some suggestion of what's going on: https://gcc.godbolt.org/z/eWY6zKz6G -- note that all the compilers reject this, because they are indeed all forming the type `Point_2<P>` eagerly, but only Clang requires it to be complete. The other compilers seem to only require the parameter type to be complete when they reach the point of looking for a conversion from an argument to the parameter type, and only look for a conversion for the second parameter if they find a conversion for the first parameter.

If you reverse the order of the parameters in `n::insert`, GCC and EDG [also reject](https://gcc.godbolt.org/z/aaffPfTjc) (but MSVC still accepts). This suggests that those compilers are bailing out early when they see there's no conversion from `s1` to `basic_ostream`, and only Clang is carrying on and checking for a conversion from `s2` to `Point_2<R>` (which is what triggers the class template instantiation and hence the error). I think that maybe MSVC is doing a "perfect match" check that allows it to pick `::insert` without considering any conversions for the other candidate (it accepts even if we force instantiation of `::insert`, so it's not just that it didn't check the call yet).
